### PR TITLE
Bug 1222135 - Implement app grouping. r=gmarty

### PR DESCRIPTION
**Pull request by Cwiiis**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=accdbcaee2c75a8e0d34e5cce0f52d70fc40f07b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8ae8f4ae22e85deb32e48ef8779ae0a665c25cd5

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=1b9224c98248fc90c0fb83ff806f1c5274c7388f

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8241d7964c03f3ed76482c994ea03c603febc708

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=71e0e732c42cbf18b52dae1dc87c44db042e4e22

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d2103d2efe62e684f61c80edd5fb3cd30fb34b6e

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a0c73ba84c9a25e9d0ef76ad18fe32aef279b464

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c6d30215cd38fa98b3a78f2cd56e3c940f75edac

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=73e1763aedc5e4b9c83cbe9e8c037db51ab54062

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c64cb401c1c44d2c6be4edab2efdf436b29840ad

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c0e46423c9cfdbac31e3e9321312aa8e65a16543

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b59a41931f992cec568ad10a4cc6d141ad9759fd

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=bdb6a40620d383fc2aba58a361c879d25e9a6c74

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=08bf74ddd7357ecc7108fefa7d1c4837ab2acab0

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d5011837f6f044061b3c1716fe19cbf77f3fea85

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f38799b5e60a65bc33c0701d0894f0608d0760de

File: apps/homescreen/js/group.js [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+  const SIZE_RATIO = 0.7;
+
+  // Maximum number of icons that can be contained in a group.
+  const MAX_CHILDREN = 6;
+
+  // Possible states of a group.
+  const COLLAPSED = 0;
+  const EXPANDING = 1;
+  const EXPANDED = 2;
+  const COLLAPSING = 3;
```
Comment (gmarty) [inline apps/homescreen/js/group.js]: What about using a pseudo enum here?
```js
const STATE = {
  COLLAPSED: 0,
  EXPANDING: 1,
  EXPANDED: 2,
  COLLAPSING: 3
};
```

File: apps/homescreen/js/apps.js [hunk 2]
```diff
@@ -401,6 +440,38 @@
...
+        if (child.localName === 'homescreen-group') {
+          for (var subContainer of child.container.children) {
+            callback(subContainer.firstElementChild,
+                     subContainer, child.container);
+          }
+        } else {
+          callback(child, container, this.icons);
+        }
+      }
+    },
```
Comment (gmarty) [inline apps/homescreen/js/apps.js]: This is very neat!

File: apps/homescreen/js/group.js [hunk 3]
```diff
@@ -0,0 +1,305 @@
+'use strict';
```
Comment (gmarty) [inline apps/homescreen/js/group.js]: nit: I'd rather have this file called `homescreen-group.js` to indicate it's a web component.
Or maybe place it in `js/webcomponent/group.js`.
Thought?

File: apps/homescreen/js/group.js [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    // Part 5.
+    var bgOffsetLeft = Math.round((targetWidth / 2) - (originalWidth / 2));
+    var bgOffsetTop = Math.round((targetHeight / 2) - (originalHeight / 2));
+    var bgLeft = Math.round(bgOffsetLeft + targetLeft) - originalLeft;
+    var bgTop = Math.round(bgOffsetTop + targetTop) - originalTop;
+    var bgTargetSize =
+      Math.sqrt(2 * Math.pow(Math.max(targetWidth, targetHeight), 2));
+    var bgScale = bgTargetSize / Math.max(originalWidth, originalHeight);
+    this.background.style.transform =
+      'translate(' + bgLeft + 'px, ' + bgTop + 'px) scale(' + bgScale + ')';
```
Comment (gmarty) [inline apps/homescreen/js/group.js]: Typically this kind of string is usually more readable when formatted with back tick `.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4f83f96424c529733f9ea43fc81d4705520aab2b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d46d27f4dc3eb7e429f125fb89aacf3cd75f200c

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4769221f61f67c046596b7a4212d32a7cd26bb25

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5ca3605cf2055b80f904f28809bb1c8a6fd564ec

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9370b971b27aeb4b02d4c15674f4f3d89a897286

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8cdc8c8edf952d9d1ba9873c6c9bc2c35a1228f5

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=33db81f8ffdea7688706019041a56a04b3a1c7bf

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ee7617a8332005a9192f7d609ad6012ce0529391

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=28b4caf66d14dbd910fdb9efe7614f6a2416fb7b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=19ae2702285e4d463c26fe18497b66d847c3b233

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ea2e084019569f87415a4e4763aae05b44a69a4c

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=840bcb3f0703285f1899386e8a984fc1f3719c8b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a6ef938bcb2c4bd73a668ed55f9a55cdf4e91b73

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=feadfb086cb6079e2c5fa91c6ad63c1ede1b4490

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=efda24f92b8e55cc7bb4c94c00214528a31350e3

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8bddbcb89a879e9e2ee8fec3297e1c2706cf95e7

File: apps/homescreen/js/apps.js [hunk 5]
```diff
@@ -477,11 +548,61 @@
...
+            if (id === groupId) {
+              parent = iconParent;
+              groupFound = true;
+            }
+          });
+
+          if (parent === this.icons && this.startupMetadata !== null) {
+            // We didn't find the group and we're still starting up, so delay
+            // adding this icon.
+            this.pendingIcons[id] = Array.slice(arguments);
```
Comment (gmarty) [inline apps/homescreen/js/apps.js]: Nit: Why not use a more readable spread operator here?
```javascript
this.pendingIcons[id] = [... arguments];
```

File: apps/homescreen/js/apps.js [hunk 6]
```diff
@@ -844,19 +985,59 @@
...
+
+      var child = element.firstElementChild;
+      return child.localName === 'homescreen-group' ?
+        'group' : child.name;
+    },
+
+    isGroup: function(element) {
+      return (element && element.firstElementChild &&
+        element.firstElementChild.localName === 'homescreen-group') ?
+        true : false;
```
Comment (gmarty) [inline apps/homescreen/js/apps.js]: Nit: Can't you simplify it to:
```javascript
return (element && element.firstElementChild &&
  element.firstElementChild.localName === 'homescreen-group');
```

File: apps/homescreen/js/apps.js [hunk 7]
```diff
@@ -885,19 +1066,35 @@
...
 
       // Disable scrolling during dragging, and display bottom-bar
       case 'drag-start':
-        console.debug('Drag-start on ' +
-                      e.detail.target.firstElementChild.name);
+        console.debug('Drag-start on ' + this.elementName(e.detail.target));
         this.dragging = true;
-        this.shouldEnterEditMode = true;
+        this.draggingGroup = this.isGroup(e.detail.target);
+        this.shouldEnterEditMode = this.openGroup ? false : true;
```
Comment (gmarty) [inline apps/homescreen/js/apps.js]: Nit: Same here, why not the following?
```javascript
this.shouldEnterEditMode = !this.openGroup;
```

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9fa51ea7a0420bba110bfac50aed025a6e286f7c

File: apps/homescreen/js/apps.js (hunk 5 above)
Comment (Cwiiis) [inline apps/homescreen/js/apps.js]: Done.

File: apps/homescreen/js/apps.js (hunk 6 above)
Comment (Cwiiis) [inline apps/homescreen/js/apps.js]: This sometimes returns null or things that aren't strictly true or false, which throws off tests. I'd like to keep it explicit.

**Pull request merged**